kore/src/Kore/Step/Simplification/ExpandAlias.hs

```diff
@@ -0,0 +1,69 @@
+{-|
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+-}
+
+module Kore.Step.Simplification.ExpandAlias
+    ( expandAlias
+    ) where
+
+import           Control.Error
+                 ( MaybeT )
+import           Control.Error.Util
+                 ( nothing )
+import qualified Data.Map as Map
+
+import           Kore.Internal.Alias
+                 ( Alias (..) )
+import           Kore.Internal.Pattern
+                 ( Pattern )
+import           Kore.Internal.TermLike
+                 ( pattern ApplyAlias_, SubstitutionVariable, TermLike,
+                 Variable, mapVariables, substitute )
+import qualified Kore.Logger as Logger
+import           Kore.Syntax.Variable
+                 ( SortedVariable (..) )
+import           Kore.Unification.Unify
+                 ( MonadUnify )
+
+expandAlias
+    :: forall variable unifier
+    .  SubstitutionVariable variable
+    => MonadUnify unifier
+    => Logger.WithLog Logger.LogMessage unifier
+    => (   TermLike variable
+        -> TermLike variable
+        -> MaybeT unifier (Pattern variable)
+       )
+    -> TermLike variable
+    -> TermLike variable
+    -> MaybeT unifier (Pattern variable)
+expandAlias recurse t1 t2 = do
+    case (expandSingleAlias t1, expandSingleAlias t2) of
+        (Nothing, Nothing) -> nothing
+        (t1', t2') -> recurse (maybe t1 id t1') (maybe t2 id t2')
+
+expandSingleAlias
+    :: SubstitutionVariable variable
+    => TermLike variable
+    -> Maybe (TermLike variable)
+expandSingleAlias =
+    \case
+        ApplyAlias_ alias children -> pure $ substituteWorker alias children
+        _ -> Nothing
+
+-- TODO(Vladimir): Check aliases such that the intersection of alias variables
+-- with the *bounded* variables in the rhs is empty (because we can't currently
+-- handle things like \mu.
+substituteWorker
+    :: SubstitutionVariable variable
+    => Alias (TermLike Variable)
+    -> [TermLike variable]
+    -> TermLike variable
+substituteWorker Alias { aliasLeft, aliasRight } children =
```

Comment:
I think we should enforce the condition that the alias definition and the pattern have the same number of arguments:

```suggestion
substituteWorker Alias { aliasLeft, aliasRight } children =
    assert (length aliasLeft == length children) $
```

I know this is checked by the pattern verifier, I just feel more comfortable enforcing the pre-conditions.

Reply:
I actually considered it, added them but then figured to check if they are needed, and indeed they are also checked in the pattern verifier so I removed it. I guess it makes sense to keep them here as well, it's not costing us anything.